File: packages/java/cnos-java/src/main/java/ai/kitsy/cnos/internal/FormulaEvaluator.java
package ai.kitsy.cnos.internal;

import ai.kitsy.cnos.CnosError;

import java.util.ArrayList;
import java.util.List;

/**
 * Evaluator for the CNOS derive expression AST.
 * Mirrors Go's evaluateDerivedFormula / evaluateNode / evaluateCall.
 */
public final class FormulaEvaluator {

    private FormulaEvaluator() {}

    @FunctionalInterface
    public interface RefResolver {
        /**
         * Resolves a config key reference.
         *
         * @param ref the key
         * @return a two-element array: {@code [value, found]} where found is Boolean
         * @throws CnosError on resolution failure
         */
        Object[] resolve(String ref) throws CnosError;
    }

    /**
     * Evaluates a parsed formula, returning the resolved value.
     *
     * @param key     the owning config key (used in error messages)
     * @param formula the parsed formula
     * @param resolver function to resolve a ref key to (value, found)
     * @return the evaluated value
     * @throws CnosError on evaluation failure or unresolvable ref
     */
    public static Object evaluate(String key, ParsedFormula formula, RefResolver resolver) throws CnosError {
        Object[] result = evaluateNode(formula.getAst(), resolver);
        Object value = result[0];
        boolean found = (Boolean) result[1];

        if ("ref".equals(formula.getAst().getKind()) && !found) {
            throw new CnosError("cnos: unable to resolve derived config key " + key
                    + " because " + formula.getAst().getPath() + " is missing");
        }
        return value;
    }

    /** Returns [value, found(Boolean)] */
    static Object[] evaluateNode(ExprNode node, RefResolver resolver) throws CnosError {
        switch (node.getKind()) {
            case "literal":
                return new Object[]{node.getValue(), Boolean.TRUE};
            case "ref":
                return resolver.resolve(node.getPath());
            case "call": {
                List<Object[]> evaluated = new ArrayList<>(node.getArgs().size());
                for (ExprNode arg : node.getArgs()) {
                    evaluated.add(evaluateNode(arg, resolver));
                }
                Object[] values = new Object[evaluated.size()];
                boolean[] flags = new boolean[evaluated.size()];
                for (int i = 0; i < evaluated.size(); i++) {
                    values[i] = evaluated.get(i)[0];
                    flags[i] = (Boolean) evaluated.get(i)[1];
                }
                return evaluateCall(node.getName(), values, flags);
            }
            default:
                throw new CnosError("cnos: unsupported derive AST node \"" + node.getKind() + "\"");
        }
    }

    private static Object[] evaluateCall(String name, Object[] values, boolean[] flags) throws CnosError {
        switch (name) {
            case "concat": {
                StringBuilder sb = new StringBuilder();
                for (Object v : values) {
                    sb.append(JsCompat.jsStringifyValue(v));
                }
                return new Object[]{sb.toString(), Boolean.TRUE};
            }
            case "coalesce": {
                for (Object v : values) {
                    if (v != null) return new Object[]{v, Boolean.TRUE};
                }
                return new Object[]{null, Boolean.TRUE};
            }
            case "when": {
                Object cond = values.length > 0 ? values[0] : null;
                Object whenTrue = values.length > 1 ? values[1] : null;
                Object whenFalse = values.length > 2 ? values[2] : null;
                return new Object[]{isTruthy(cond) ? whenTrue : whenFalse, Boolean.TRUE};
            }
            case "exists": {
                if (values.length == 0) return new Object[]{Boolean.FALSE, Boolean.TRUE};
                boolean exists = flags[0] && values[0] != null;
                return new Object[]{exists, Boolean.TRUE};
            }
            case "eq": {
                Object left = values.length > 0 ? values[0] : null;
                Object right = values.length > 1 ? values[1] : null;
                return new Object[]{JsCompat.jsStrictEqual(left, right), Boolean.TRUE};
            }
            case "ne": {
                Object left = values.length > 0 ? values[0] : null;
                Object right = values.length > 1 ? values[1] : null;
                return new Object[]{!JsCompat.jsStrictEqual(left, right), Boolean.TRUE};
            }
            default:
                throw new CnosError("cnos: unknown derive function: " + name);
        }
    }

    static boolean isTruthy(Object value) {
        if (value == null) return false;
        if (value instanceof Boolean) return (Boolean) value;
        if (value instanceof String) return !((String) value).isEmpty();
        if (value instanceof Double) return (Double) value != 0.0;
        if (value instanceof Float) return (Float) value != 0.0f;
        if (value instanceof Long) return (Long) value != 0L;
        if (value instanceof Integer) return (Integer) value != 0;
        if (value instanceof Number) return ((Number) value).doubleValue() != 0;
        return true; // non-null objects are truthy
    }
}

File: packages/csharp/Kitsy.Cnos/Internal/FormulaEvaluator.cs
using System;
using System.Collections.Generic;
using System.Text;
using Kitsy.Cnos;

namespace Kitsy.Cnos.Internal
{
    internal static class FormulaEvaluator
    {
        public delegate (object? Value, bool Found) RefResolver(string key);

        /// <summary>Evaluates a parsed formula. Throws CnosError on failure.</summary>
        public static object? Evaluate(string key, ParsedFormula formula, RefResolver resolveRef)
        {
            var (value, found) = EvaluateNode(formula.Ast, resolveRef);
            if (formula.Ast.Kind == "ref" && !found)
                throw new CnosError($"cnos: unable to resolve derived config key {key} because {formula.Ast.Path} is missing");
            return value;
        }

        private static (object? Value, bool Found) EvaluateNode(ExprNode node, RefResolver resolveRef)
        {
            return node.Kind switch
            {
                "literal" => (node.Value, true),
                "ref" => resolveRef(node.Path),
                "call" => EvaluateCall(node.Name, node.Args, resolveRef),
                _ => throw new CnosError($"cnos: unsupported derive AST node \"{node.Kind}\""),
            };
        }

        private static (object? Value, bool Found) EvaluateCall(string name, IReadOnlyList<ExprNode> args, RefResolver resolveRef)
        {
            var values = new object?[args.Count];
            var flags = new bool[args.Count];
            for (int i = 0; i < args.Count; i++)
            {
                var (v, f) = EvaluateNode(args[i], resolveRef);
                values[i] = v;
                flags[i] = f;
            }

            return name switch
            {
                "concat" => EvalConcat(values),
                "coalesce" => EvalCoalesce(values),
                "when" => EvalWhen(values),
                "exists" => EvalExists(values, flags),
                "eq" => (JsCompat.JsStrictEqual(At(values, 0), At(values, 1)), true),
                "ne" => (!JsCompat.JsStrictEqual(At(values, 0), At(values, 1)), true),
                _ => throw new CnosError($"cnos: unknown derive function: {name}"),
            };
        }

        private static (object? Value, bool Found) EvalConcat(object?[] values)
        {
            var sb = new StringBuilder();
            foreach (var v in values) sb.Append(JsCompat.JsStringifyValue(v));
            return (sb.ToString(), true);
        }

        private static (object? Value, bool Found) EvalCoalesce(object?[] values)
        {
            foreach (var v in values) if (v != null) return (v, true);
            return (null, true);
        }

        private static (object? Value, bool Found) EvalWhen(object?[] values)
        {
            object? whenTrue = values.Length > 1 ? values[1] : null;
            object? whenFalse = values.Length > 2 ? values[2] : null;
            return (JsCompat.IsTruthy(At(values, 0)) ? whenTrue : whenFalse, true);
        }

        private static (object? Value, bool Found) EvalExists(object?[] values, bool[] flags) =>
            (values.Length > 0 && flags[0] && values[0] != null, true);

        private static object? At(object?[] values, int index) =>
            index >= 0 && index < values.Length ? values[index] : null;
    }
}
